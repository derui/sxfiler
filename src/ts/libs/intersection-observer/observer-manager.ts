// define application-global intersection observer management.
import uuid from "uuid";
import "intersection-observer";

export type RootId = string;
export type IntersectionChandedHandler = (entry: IntersectionObserverEntry) => void;

export interface IntersectionTarget {
  element: Element;
  handler: IntersectionChandedHandler;
}

interface Observer {
  instance: IntersectionObserver | undefined;
  targets: Map<Element, IntersectionTarget>;
}

const intersectionObserverMap = new Map<RootId, Observer>();

// get observer binded rootId
function getObserver(rootId: RootId): Observer | undefined {
  return intersectionObserverMap.get(rootId);
}

export function observingTargets(rootId: RootId): IntersectionTarget[] | undefined {
  const observer = intersectionObserverMap.get(rootId);

  if (!observer) {
    return undefined;
  }

  return Array.from(observer.targets.values());
}

// observe the element by root
export function observe(rootId: RootId, target: IntersectionTarget) {
  const observer = getObserver(rootId);

  if (!observer) {
    throw new Error(`Not found root instance by ${rootId}, try "createRoot" first.`);
  }

  observer.targets.set(target.element, target);

  const { instance } = observer;
  if (instance) {
    instance.observe(target.element);
  }
}

// unobserve the element from root
export function unobserve(target: IntersectionTarget): boolean {
  let deleted = false;

  for (let entry of intersectionObserverMap.entries()) {
    let [_, observer] = entry;
    if (observer.instance) {
      observer.instance.unobserve(target.element);
    }
    deleted = observer.targets.delete(target.element) || deleted;
  }

  return deleted;
}

// create and register new root
export function createRoot(): RootId {
  const id = uuid.v4();

  intersectionObserverMap.set(id, { instance: undefined, targets: new Map() });

  return id;
}

export function deleteRoot(rootId: RootId): boolean {
  const observer = getObserver(rootId);

  if (!observer) {
    return false;
  }

  intersectionObserverMap.delete(rootId);

  const { instance, targets } = observer;
  if (instance) {
    targets.forEach(e => instance.unobserve(e.element));
    instance.disconnect();
  }

  return true;
}

function callback(rootId: RootId): (entries: IntersectionObserverEntry[]) => void {
  return entries => {
    const observer = getObserver(rootId);

    if (!observer) {
      throw new Error(`Not found root instance of ${rootId}`);
    }

    entries.forEach(entry => {
      const target = observer.targets.get(entry.target);
      if (target) {
        target.handler(entry);
      }
    });
  };
}

export function registerObserverRoot(rootId: RootId, option: IntersectionObserverInit): IntersectionObserver {
  const observer = getObserver(rootId);

  if (!observer) {
    throw new Error(`Not found root ${rootId}`);
  }

  const { instance, targets } = observer;
  if (observer && instance && instance.root !== option.root) {
    targets.forEach(e => instance.unobserve(e.element));
    instance.disconnect();
    observer.instance = undefined;
  }

  const newInstance = new IntersectionObserver(callback(rootId), option);
  observer.instance = newInstance;
  targets.forEach(e => newInstance.observe(e.element));
  return newInstance;
}
