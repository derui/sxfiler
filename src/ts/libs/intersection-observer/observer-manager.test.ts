import * as observer from "./observer-manager";

afterEach(() => {
  jest.restoreAllMocks();
});

describe("Observer Manager", () => {
  it("create and delete root", () => {
    const id = observer.createRoot();

    expect(observer.deleteRoot(id)).toBeTruthy();
  });

  it("should be able to observe and unobserve the target before register target", () => {
    const id = observer.createRoot();

    const target = {
      element: document.createElement("span"),
      handler: () => {},
    };
    observer.observe(id, target);

    expect(observer.unobserve(target)).toEqual(true);
  });

  it("should unobserve from all observer", () => {
    const id = observer.createRoot();
    const id2 = observer.createRoot();

    const target = {
      element: document.createElement("span"),
      handler: () => {},
    };
    observer.observe(id, target);
    observer.observe(id2, target);

    expect(observer.unobserve(target)).toEqual(true);
    expect(observer.observingTargets(id)).toHaveLength(0);
    expect(observer.observingTargets(id2)).toHaveLength(0);
  });

  it("should throw error if root is not found", () => {
    const target = {
      element: document.createElement("span"),
      handler: () => {},
    };

    expect(() => observer.observe("id", target)).toThrow();
  });

  it("register observer root with option", () => {
    const original = IntersectionObserver;
    const mockInstance = jest.fn();
    IntersectionObserver = mockInstance;

    const id = observer.createRoot();
    const option = {
      root: document.createElement("span"),
    };
    observer.registerObserverRoot(id, option);

    expect(mockInstance.mock.instances[0]).not.toBeNull();

    IntersectionObserver = original;
  });

  it("throws if rootMargin is invalid", () => {
    const id = observer.createRoot();
    const option = {
      root: document.createElement("span"),
      rootMargin: "10",
    };
    expect(() => observer.registerObserverRoot(id, option)).toThrowErrorMatchingSnapshot();
  });

  it("unregister existance instance if register new root ", () => {
    const id = observer.createRoot();
    const option = {
      root: document.createElement("span"),
    };
    const firstObserver = observer.registerObserverRoot(id, option);
    const spy = jest.spyOn(firstObserver, "disconnect");
    const secondObserver = observer.registerObserverRoot(id, {
      root: document.createElement("div"),
    });

    expect(spy.mock.calls).toHaveLength(1);
    expect(firstObserver).not.toStrictEqual(secondObserver);
  });

  it("unregister observing element if register new root ", () => {
    const id = observer.createRoot();
    const target = {
      element: document.createElement("a"),
      handler: () => {},
    };
    const option = {
      root: document.createElement("span"),
    };
    const firstObserver = observer.registerObserverRoot(id, option);
    observer.observe(id, target);

    const spy = jest.spyOn(firstObserver, "disconnect");
    const spyUnobserve = jest.spyOn(firstObserver, "unobserve");
    const secondObserver = observer.registerObserverRoot(id, {
      root: document.createElement("div"),
    });

    expect(spy.mock.calls).toHaveLength(1);
    expect(spyUnobserve.mock.calls[0][0]).toEqual(target.element);
    expect(firstObserver).not.toStrictEqual(secondObserver);
  });
});
