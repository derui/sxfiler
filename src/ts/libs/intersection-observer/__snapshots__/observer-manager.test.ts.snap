// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Observer Manager throws if rootMargin is invalid 1`] = `"rootMargin must be specified in pixels or percent"`;
